src/Microsoft.AspNet.Mvc.TagHelpers/InputTagHelper.cs

```diff
@@ -389,15 +390,14 @@ private string GetInputType(ModelExplorer modelExplorer, out string inputTypeHin
         // A variant of TemplateRenderer.GetViewNames(). Main change relates to bool? handling.
         private static IEnumerable<string> GetInputTypeHints(ModelExplorer modelExplorer)
         {
-            var inputTypeHints = new string[]
+            if (!string.IsNullOrEmpty(modelExplorer.Metadata.TemplateHint))
             {
-                modelExplorer.Metadata.TemplateHint,
-                modelExplorer.Metadata.DataTypeName,
-            };
+                yield return modelExplorer.Metadata.TemplateHint;
+            }
 
-            foreach (string inputTypeHint in inputTypeHints.Where(s => !string.IsNullOrEmpty(s)))
+            if (!string.IsNullOrEmpty(modelExplorer.Metadata.DataTypeName))
             {
-                yield return inputTypeHint;
+                yield return modelExplorer.Metadata.DataTypeName;
```

Comment:
How expensive is this state machine? Would we be better off returning a `List<string>`?